fv/policysync_test.go

```diff
@@ -0,0 +1,511 @@
+			os.MkdirAll(hostWlDir, 0777)
+			return hostWlDir, "/var/run/calico/" + dirName
+		}
+
+		sendCreate := func(wl *workload.Workload) (*mgmtintf_v1.Response, error) {
+			dirName := dirNameForWorkload(wl)
+
+			resp, err := mgmtClient.WorkloadAdded(&mgmtintf_v1.WorkloadInfo{
+				Attrs: &mgmtintf_v1.WorkloadInfo_WorkloadAttributes{
+					Uid:       wl.WorkloadEndpoint.Spec.Pod,
+					Namespace: "fv",
+					Workload:  wl.WorkloadEndpoint.Spec.Pod,
+				},
+				Workloadpath: dirName,
+			})
+			log.WithField("response", resp).Info("WorkloadAdded response")
+			return resp, err
+		}
+
+		Context("after creating a client for each workload", func() {
+			BeforeEach(func() {
+				for i, wl := range w {
+					// Create the workload directory, this would normally be the responsibility of the
+					// flex volume driver.
+					hostWlDir, containerWlDir := createWorkloadDirectory(wl)
+					hostWlSocketPath[i] = hostWlDir + "/policysync.sock"
+					containerWlSocketPath[i] = containerWlDir + "/policysync.sock"
+
+					// Tell Felix about the new directory.
+					_, err := sendCreate(wl)
+					Expect(err).NotTo(HaveOccurred())
+				}
+			})
+
+			It("felix should create the workload socket", func() {
+				for _, p := range hostWlSocketPath {
+					Eventually(p).Should(BeAnExistingFile())
+				}
+			})
+
+			Context("with open workload connections", func() {
+
+				// Then connect to it.
+				var (
+					wlConn   [3]*grpc.ClientConn
+					wlClient [3]proto.PolicySyncClient
+					cancel   context.CancelFunc
+					ctx      context.Context
+					err      error
+				)
+
+				createWorkloadConn := func(i int) (*grpc.ClientConn, proto.PolicySyncClient) {
+					var opts []grpc.DialOption
+					opts = append(opts, grpc.WithInsecure())
+					opts = append(opts, grpc.WithDialer(unixDialer))
+					var conn *grpc.ClientConn
+					conn, err = grpc.Dial(hostWlSocketPath[i], opts...)
+					Expect(err).NotTo(HaveOccurred())
+					wlClient := proto.NewPolicySyncClient(conn)
+					return conn, wlClient
+				}
+
+				BeforeEach(func() {
+					ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
+
+					for i := range w {
+						// Use the fact that anything we exec inside the Felix container runs as root to fix the
+						// permissions on the socket so the test process can connect.
+						Eventually(hostWlSocketPath[i]).Should(BeAnExistingFile())
+						felix.Exec("chmod", "a+rw", containerWlSocketPath[i])
+						wlConn[i], wlClient[i] = createWorkloadConn(i)
+					}
+				})
+
+				AfterEach(func() {
+					if cancel != nil {
+						cancel()
+					}
+				})
+
+				Context("with mock clients syncing", func() {
+					var (
+						mockWlClient [3]*mockWorkloadClient
+						mockWlCancel [3]context.CancelFunc
+					)
+
+					BeforeEach(func() {
+						for i := range w {
+							client := newMockWorkloadClient(fmt.Sprintf("workload-%d", i))
+							var wlCtx context.Context
+							wlCtx, mockWlCancel[i] = context.WithCancel(ctx)
```

Comment:
These Cancel-able contexts are a great idea, but there are no tests that invoke them.

Reply:
Yes, I'd intended to use them to interrupt clients but the gRPC client API doesn't take a context so they were less useful than I'd hoped.  They only stop the loop after the next `Recv()`.

Reply:
I don't understand; `Sync` accepts a context:
```
Sync(ctx context.Context, in *SyncRequest, opts ...grpc.CallOption) (PolicySync_SyncClient, error)
```

Are you saying that cancelling only takes effect after the next call to `Recv()`?